python/tk_nuke/__init__.py

```diff
@@ -144,56 +142,59 @@ def tank_startup_node_callback():
     
     Carefully manage exceptions here so that a bug in Tank never
     interrupts the normal workflows in Nuke.    
-    """    
+    """
     try:    
-        if nuke.root().name() == "Root":
-            # file->new
-            # base it on the context we 'inherited' from the prev session
-            # get the context from the previous session - this is helpful if user does file->new
-            project_root = os.environ.get("TANK_NUKE_ENGINE_INIT_PROJECT_ROOT")
-            tk = tank.Tank(project_root)
-            
-            ctx_str = os.environ.get("TANK_NUKE_ENGINE_INIT_CONTEXT")
-            if ctx_str:
-                try:
-                    new_ctx = tank.context.deserialize(ctx_str)
-                except:
-                    new_ctx = tk.context_empty()
-            else:
+
+        # first load up in our previous context so we can start the engine an work out if automatic context switching
+        # is enabled, and if it is we should then switch to the correct environment.
+        # base it on the context we 'inherited' from the prev session
+        # get the context from the previous session
+        project_root = os.environ.get("TANK_NUKE_ENGINE_INIT_PROJECT_ROOT")
+        tk = tank.Tank(project_root)
+
+        ctx_str = os.environ.get("TANK_NUKE_ENGINE_INIT_CONTEXT")
+        if ctx_str:
+            try:
+                new_ctx = tank.context.deserialize(ctx_str)
+            except Exception:
                 new_ctx = tk.context_empty()
-    
         else:
+            new_ctx = tk.context_empty()
+
+        # now restart the engine with the new context
+        __engine_refresh(tk, new_ctx)
+
+        # If we have opened a file then we should check if automatic context switching is enabled and change if possible
+        engine = tank.platform.current_engine()
+        if nuke.root().name() != "Root" and engine.get_setting("automatic_context_switch"):
             # file->open
             file_name = nuke.root().name()
-            
+
             try:
                 tk = tank.tank_from_path(file_name)
```

Comment:
We can maybe fetch the `tank` object from the current engine instead of creating it in each case, what do you think ?